File: leetcode/design/ImplementQueueUsingStacks.java
package design;

public class ImplementQueueUsingStacks {
    /* 232
     * 把s1的数字导到s2里 这样s2当中 顺序就反过来了 变成了FIFO 也就是queue的顺序
     * s2的肯定要先用完/pop完 因为顺序在前 这时候push先push到s1里 一会儿s2用完了 再把s1的刚加的导过去
     */
    private Stack<Integer> s1, s2;
    public MyQueue() {
        s1 = new Stack<>();
        s2 = new Stack<>();
    }
    
    public void push(int x) {
        s1.push(x);
    }
    
    public int pop() {
        if(s2.isEmpty()) {
            while(!s1.isEmpty()) {
                s2.push(s1.pop());
            }
        }
        return s2.pop();
    }
    
    public int peek() {
        if(s2.isEmpty()) {
            while(!s1.isEmpty()) {
                s2.push(s1.pop());
            }
        }
        return s2.peek();
    }
    
    public boolean empty() {
        return s1.size() == 0 && s2.size() == 0;
    }
}

File: c++/stack/implementQueueUsingStacks.cpp
using namespace std;
/**
 * 232. Implement Queue using Stacks
 * 用两个stack倒一下 push正常push 拿的时候倒到out stack里 这样顺序就反过来了
 * 后面接着push先放在in里 直到out拿空则再次翻转顺序
*/
class MyQueue {
private:
    stack<int> in, out;
public:
    MyQueue() {
        
    }
    
    void push(int x) {
        in.push(x);
    }
    
    int pop() {
        in_and_out();
        int top = out.top();
        out.pop();
        return top;
    }

    void in_and_out(){
        if(out.empty()) {
            while(!in.empty()) {
                int x = in.top();
                in.pop();
                out.push(x);
            }
        }
    }
    
    int peek() {
        in_and_out();
        return out.top();
    }
    
    bool empty() {
        return in.empty() && out.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */